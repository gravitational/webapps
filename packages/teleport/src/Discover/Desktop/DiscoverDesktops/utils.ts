import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
