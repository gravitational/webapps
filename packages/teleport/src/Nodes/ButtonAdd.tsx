/*
Copyright 2020 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { ButtonPrimary } from 'design';

const docUrl =
  'https://gravitational.com/teleport/docs/admin-guide/#adding-nodes-to-the-cluster';

export default function ButtonAdd(props: Props) {
  const { isEnterprise, canCreate, isLeafCluster, ...rest } = props;

  if (!isEnterprise) {
    return (
      <ButtonPrimary
        {...rest}
        width="240px"
        onClick={() => null}
        as="a"
        target="_blank"
        href={docUrl}
      >
        View documentation
      </ButtonPrimary>
    );
  }

  if (canCreate) {
    const title = isLeafCluster
      ? 'Adding a server to a leaf cluster is not supported'
      : 'Add a server to the root cluster';

    return (
      <ButtonPrimary
        title={title}
        disabled={isLeafCluster}
        width="240px"
        {...props}
      >
        Add Server
      </ButtonPrimary>
    );
  }

  return null;
}

type Props = {
  isLeafCluster: boolean;
  isEnterprise: boolean;
  canCreate: boolean;
  onClick?: () => void;
  mb?: string;
  mx?: string;
  width?: string;
  kind?: string;
};
