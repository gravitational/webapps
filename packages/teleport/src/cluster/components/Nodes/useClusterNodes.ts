/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { useState, useEffect, useCallback } from 'react';
import { useAttempt } from 'shared/hooks';
import { useStore } from 'shared/libs/stores';
import TeleportContext from 'teleport/teleportContext';
import cfg from 'teleport/config';
import { Node } from 'teleport/services/nodes';

export default function useClusterNodes(teleCtx: TeleportContext) {
  const storeUser = useStore(teleCtx.storeUser);
  const [nodes, setNodes] = useState<Node[]>([]);
  const [attempt, attemptActions] = useAttempt({ isProcessing: true });
  const logins = storeUser.getLogins();

  useEffect(() => {
    attemptActions.do(() => teleCtx.nodeService.fetchNodes().then(setNodes));
  }, []);

  const getNodeLoginOptions = useCallback(
    (serverId: string) => makeOptions(serverId, logins),
    [logins]
  );

  const openNewTab = (url: string) => {
    const element = document.createElement('a');
    element.setAttribute('href', `${url}`);
    // works in ie11
    element.setAttribute('target', `_blank`);
    element.style.display = 'none';
    document.body.appendChild(element);
    element.click();
    document.body.removeChild(element);
  };

  const startSshSession = (login: string, serverId: string) => {
    const url = cfg.getSshConnectRoute({
      serverId,
      login,
    });

    openNewTab(url);
  };

  return {
    attempt,
    nodes,
    getNodeLoginOptions,
    startSshSession,
  };
}

function makeOptions(serverId = '', logins = [] as string[]) {
  return logins.map(login => {
    const url = cfg.getSshConnectRoute({
      serverId,
      login,
    });

    return {
      login,
      url,
    };
  });
}
