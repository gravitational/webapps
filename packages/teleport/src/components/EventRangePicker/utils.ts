import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
