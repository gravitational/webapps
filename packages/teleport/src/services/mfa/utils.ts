/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
