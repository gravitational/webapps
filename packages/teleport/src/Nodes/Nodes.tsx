/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Indicator, Box, Flex } from 'design';
import { Danger } from 'design/Alert';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import QuickLaunch from 'teleport/components/QuickLaunch';
import InputSearch from 'teleport/components/InputSearch';
import NodeList from 'teleport/components/NodeList';
import useTeleport from 'teleport/useTeleport';
import useStickyClusterId from 'teleport/useStickyClusterId';
import useNodes, { State } from './useNodes';
import AddNode from './AddNode';
import ButtonAdd from './ButtonAdd';

export default function Container() {
  const teleCtx = useTeleport();
  const { clusterId } = useStickyClusterId();
  const state = useNodes(teleCtx, clusterId);
  return <Nodes {...state} />;
}

export function Nodes(props: State) {
  const {
    nodes,
    searchValue,
    setSearchValue,
    getNodeLoginOptions,
    startSshSession,
    attempt,
    showAddNode,
    canCreate,
    hideAddNode,
    isAddNodeVisible,
    isEnterprise,
  } = props;

  function onLoginSelect(e: React.MouseEvent, login: string, serverId: string) {
    e.preventDefault();
    startSshSession(login, serverId);
  }

  function onSshEnter(login: string, serverId: string) {
    startSshSession(login, serverId);
  }

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center" justifyContent="space-between">
        <FeatureHeaderTitle>Servers</FeatureHeaderTitle>
        <ButtonAdd
          isEnterprise={isEnterprise}
          canCreate={canCreate}
          onClick={showAddNode}
        />
      </FeatureHeader>
      <Flex
        mb={4}
        alignItems="center"
        flex="0 0 auto"
        justifyContent="space-between"
      >
        <InputSearch mr="3" onChange={setSearchValue} />
        <QuickLaunch width="280px" onPress={onSshEnter} />
      </Flex>
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'success' && (
        <NodeList
          nodes={nodes}
          searchValue={searchValue}
          onLoginMenuOpen={getNodeLoginOptions}
          onLoginSelect={onLoginSelect}
        />
      )}
      {isAddNodeVisible && <AddNode onClose={hideAddNode} />}
    </FeatureBox>
  );
}
