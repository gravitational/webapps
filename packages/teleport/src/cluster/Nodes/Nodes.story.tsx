/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Nodes } from './Nodes';

type PropTypes = Parameters<typeof Nodes>[0];

export default {
  title: 'TeleportCluster/Nodes',
};

export function Loaded() {
  return render({ isSuccess: true });
}

export function Loading() {
  return render({ isProcessing: true });
}

export function Failed() {
  return render({ isFailed: true, message: 'server error' });
}

function render(attemptOptions: Partial<PropTypes['attempt']>) {
  const attempt = {
    isProcessing: false,
    isSuccess: false,
    isFailed: false,
    message: '',
    ...attemptOptions,
  };

  return (
    <Nodes
      attempt={attempt}
      startSshSession={() => null}
      getNodeLoginOptions={() => [{ login: 'root', url: 'fd' }]}
      nodes={nodes}
    />
  );
}

const nodes = [
  {
    tunnel: false,
    id: '104',
    clusterId: 'one',
    hostname: 'fujedu',
    addr: '172.10.1.20:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: false,
    id: '170',
    clusterId: 'one',
    hostname: 'facuzguv',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: false,
    id: '192',
    clusterId: 'one',
    hostname: 'duzsevkig',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: false,
    id: '64',
    clusterId: 'one',
    hostname: 'kuhinur',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: false,
    id: '81',
    clusterId: 'one',
    hostname: 'zebpecda',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: true,
    id: '81',
    clusterId: 'one',
    hostname: 'zebpecda',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
  {
    tunnel: true,
    id: '81',
    clusterId: 'one',
    hostname: 'zebpecda',
    addr: '172.10.1.1:3022',
    tags: [
      { name: 'cluster', value: 'one' },
      { name: 'kernel', value: '4.15.0-51-generic' },
    ],
  },
];
