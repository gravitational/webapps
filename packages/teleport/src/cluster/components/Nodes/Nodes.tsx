/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import * as Cards from 'design/CardError';
import { Indicator, Box, Flex } from 'design';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import QuickLaunch from 'teleport/components/QuickLaunch';
import { useTeleport } from 'teleport/teleportContextProvider';
import NodeList from 'teleport/components/NodeList';
import useClusterNodes from './useClusterNodes';

export default function ClusterNodes() {
  const teleCtx = useTeleport();
  const state = useClusterNodes(teleCtx);
  return <Nodes {...state} />;
}

export function Nodes({
  nodes,
  getNodeLoginOptions,
  startSshSession,
  attempt,
}: NodesProp) {
  if (attempt.isFailed) {
    return <Cards.Failed alignSelf="baseline" message={attempt.message} />;
  }

  function onLoginSelect(e: React.MouseEvent, login: string, serverId: string) {
    e.preventDefault();
    startSshSession(login, serverId);
  }

  function onQuickLaunchEnter(login: string, serverId: string) {
    startSshSession(login, serverId);
  }

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center" justifyContent="space-between">
        <FeatureHeaderTitle mr="5">Nodes</FeatureHeaderTitle>
        <QuickLaunch
          as={Flex}
          autoFocus={false}
          alignItems="center"
          inputProps={{
            style: {
              height: '30px',
              fontSize: '12px',
            },
          }}
          labelProps={{
            mr: 3,
            mb: 0,
            style: { whiteSpace: 'nowrap', width: 'auto' },
          }}
          onPress={onQuickLaunchEnter}
        />
      </FeatureHeader>
      {attempt.isProcessing && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.isSuccess && (
        <NodeList
          onLoginMenuOpen={getNodeLoginOptions}
          nodes={nodes}
          onLoginSelect={onLoginSelect}
        />
      )}
    </FeatureBox>
  );
}

type NodesProp = ReturnType<typeof useClusterNodes>;
