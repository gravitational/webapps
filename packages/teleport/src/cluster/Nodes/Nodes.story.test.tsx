import React from 'react';
import { Loaded, Failed } from './Nodes.story';
import { render } from 'design/utils/testing';

test('loaded', () => {
  const { container } = render(<Loaded />);
  expect(container.firstChild).toMatchSnapshot();
});

test('failed', () => {
  const { container } = render(<Failed />);
  expect(container.firstChild).toMatchSnapshot();
});
