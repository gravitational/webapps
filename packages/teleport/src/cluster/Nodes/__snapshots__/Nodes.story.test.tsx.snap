// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failed 1`] = `
.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 2px;
  box-sizing: border-box;
  box-shadow: 0 1px 4px rgba(0,0,0,.24);
  margin: 0 0 24px 0;
  min-height: 40px;
  padding: 8px 16px;
  overflow: auto;
  word-break: break-all;
  line-height: 1.5;
  margin-bottom: 24px;
  margin-top: 8px;
  background: #f50057;
  color: #FFFFFF;
}

.c2 a {
  color: #FFFFFF;
}

.c0 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  padding: 32px;
  width: 540px;
  color: #324148;
  background-color: #FFFFFF;
  align-self: baseline;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c1 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 28px;
  line-height: 32px;
  margin: 0px;
  margin-bottom: 24px;
  text-align: center;
}

<div
  class="sc-AxirZ c0"
  color="text.onLight"
  width="540px"
>
  <div
    class="c1"
  >
    Internal Error
  </div>
  <div
    class="c2"
    kind="danger"
  >
    server error
  </div>
   
</div>
`;

exports[`loaded 1`] = `
.c4 {
  box-sizing: border-box;
}

.c18 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #2C3A73;
  border: 1px solid #1C254D;
  opacity: .87;
  color: rgba(255,255,255,0.87);
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  height: 24px;
}

.c18:active {
  opacity: 0.56;
}

.c18:hover,
.c18:focus {
  background: #2C3A73;
  border: 1px solid rgba(255,255,255,0.1);
  opacity: 1;
}

.c18:active {
  opacity: 0.24;
}

.c18:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c13 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
  font-size: 16px;
}

.c16 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c19 {
  display: inline-block;
  transition: color .3s;
  margin-left: 8px;
  margin-right: -8px;
  color: rgba(255,255,255,0.56);
  font-size: 14px;
}

.c17 {
  box-sizing: border-box;
  border-radius: 100px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  min-height: 16px;
  line-height: 1.4;
  font-size: 10px;
  font-weight: 500;
  padding: 0 8px;
  background-color: #111B48;
  color: rgba(255,255,255,0.87);
  margin-bottom: 4px;
  margin-right: 4px;
}

.c5 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 0px;
  margin-right: 16px;
}

.c10 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: #FFFFFF;
}

.c3 {
  box-sizing: border-box;
  display: flex;
  align-items: center;
}

.c7 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  margin-right: 32px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c6 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: rgba(255,255,255,0.87);
  background-color: #222C59;
  width: 200px;
  height: 34px;
  background: #222C59;
  border: 1px solid #111B48;
  border-radius: 4px;
  border-color: rgba(255,255,255,0.24);
  font-size: 14px;
  font-family: Ubuntu2,-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
}

.c6::-ms-clear {
  display: none;
}

.c6::placeholder {
  opacity: 0.4;
}

.c6:hover,
.c6:focus,
.c6:active {
  color: rgba(255,255,255,0.87);
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
}

.c6::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 14px;
}

.c14 {
  background: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-collapse: collapse;
  border-spacing: 0;
  border-radius: 8px;
  font-size: 12px;
  width: 100%;
}

.c14 > thead > tr > th,
.c14 > tbody > tr > th,
.c14 > tfoot > tr > th,
.c14 > thead > tr > td,
.c14 > tbody > tr > td,
.c14 > tfoot > tr > td {
  padding: 8px 8px;
  vertical-align: middle;
}

.c14 > thead > tr > th:first-child,
.c14 > tbody > tr > th:first-child,
.c14 > tfoot > tr > th:first-child,
.c14 > thead > tr > td:first-child,
.c14 > tbody > tr > td:first-child,
.c14 > tfoot > tr > td:first-child {
  padding-left: 24px;
}

.c14 > thead > tr > th:last-child,
.c14 > tbody > tr > th:last-child,
.c14 > tfoot > tr > th:last-child,
.c14 > thead > tr > td:last-child,
.c14 > tbody > tr > td:last-child,
.c14 > tfoot > tr > td:last-child {
  padding-right: 24px;
}

.c14 > thead > tr > th {
  background: #111B48;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 10px;
  font-weight: 400;
  padding-bottom: 0;
  padding-top: 0;
  text-align: left;
  opacity: 0.75;
  text-transform: uppercase;
  white-space: nowrap;
}

.c14 > thead > tr > th .c12 {
  font-weight: bold;
  font-size: 8px;
  margin-left: 8px;
}

.c14 > tbody > tr > td {
  color: rgba(255,255,255,0.87);
  line-height: 16px;
}

.c14 tbody tr {
  border-bottom: 1px solid #1C254D;
}

.c14 tbody tr:hover {
  background-color: rgb(37,49,98);
}

.c11 {
  box-sizing: border-box;
  display: flex;
}

.c11 button {
  background: none;
  border: none;
  border-radius: 200px;
  cursor: pointer;
  padding: 0;
  margin: 0 0 0 8px;
  outline: none;
  transition: all 0.3s;
  text-align: center;
}

.c11 button:hover,
.c11 button:focus {
  background: #1C254D;
}

.c11 button:hover .c12,
.c11 button:focus .c12 {
  opacity: 1;
}

.c11 button .c12 {
  opacity: 0.4;
  font-size: 20px;
  transition: all 0.3s;
}

.c9 {
  padding: 8px 24px;
  display: flex;
  height: 24px;
  flex-shrink: 0;
  align-items: center;
  justify-content: space-between;
  background: #222C59;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c8 {
  box-sizing: border-box;
  font-size: 12px;
  min-width: 200px;
  outline: none;
  border: none;
  border-radius: 200px;
  height: 32px;
  transition: all .2s;
  background: #222C59;
  margin-right: 16px;
  padding-left: 16px;
  padding-right: 16px;
  color: rgba(255,255,255,0.87);
  height: 30px;
}

.c8:hover {
  background: #2C3A73;
}

.c8:focus,
.c8:active {
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  color: rgba(255,255,255,0.87);
}

.c8::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 12px;
}

.c15 > tbody > tr > td {
  vertical-align: baseline;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      Nodes
    </div>
    <div
      class="c3 sc-AxirZ c4"
    >
      <label
        class="c5"
        font-size="0"
        style="white-space: nowrap; width: auto;"
      >
        Quick Launch
      </label>
      <input
        class="sc-fzozJi c6"
        color="text.primary"
        height="34px"
        placeholder="login@host"
        style="height: 30px; font-size: 12px;"
        width="200px"
      />
    </div>
  </div>
  <div>
    <div
      class="c7"
    >
      <input
        class="c8"
        color="text.primary"
        height="30px"
        placeholder="SEARCH..."
        value=""
      />
    </div>
    <div>
      <nav
        class="c9"
      >
        <div
          class="c10"
          color="primary.contrastText"
        >
          SHOWING 
          <strong>
            1
          </strong>
           - 
          <strong>
            7
          </strong>
           
          of 
          <strong>
            7
          </strong>
        </div>
        <div
          class="c11"
        >
          <button
            disabled=""
            title="Previous Page"
          >
            <span
              class="c12 c13 icon icon-arrow-left-circle "
              color="light"
              font-size="3"
            />
          </button>
          <button
            disabled=""
            title="Next Page"
          >
            <span
              class="c12 c13 icon icon-arrow-right-circle "
              color="light"
              font-size="3"
            />
          </button>
        </div>
      </nav>
      <table
        class="c14 c15"
      >
        <thead>
          <tr>
            <th>
              <a>
                Hostname
                <span
                  class="c12 c16 icon icon-chevron-down "
                  color="light"
                />
              </a>
            </th>
            <th>
              <a>
                Address
                <span
                  class="c12 c16 icon icon-chevrons-expand-vertical "
                  color="light"
                />
              </a>
            </th>
            <th>
              Labels
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              duzsevkig
            </td>
            <td>
              172.10.1.1:3022
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              facuzguv
            </td>
            <td>
              172.10.1.1:3022
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              fujedu
            </td>
            <td>
              172.10.1.20:3022
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              kuhinur
            </td>
            <td>
              172.10.1.1:3022
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              zebpecda
            </td>
            <td>
              172.10.1.1:3022
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              zebpecda
            </td>
            <td>
              <span
                style="cursor: default;"
                title="This node is connected to cluster through reverse tunnel"
              >
                ⟵ tunnel
              </span>
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
          <tr>
            <td>
              zebpecda
            </td>
            <td>
              <span
                style="cursor: default;"
                title="This node is connected to cluster through reverse tunnel"
              >
                ⟵ tunnel
              </span>
            </td>
            <td>
              <div
                class="c17"
                kind="secondary"
              >
                cluster: one
              </div>
              <div
                class="c17"
                kind="secondary"
              >
                kernel: 4.15.0-51-generic
              </div>
            </td>
            <td
              align="right"
            >
              <button
                class="c18"
                height="24px"
                kind="border"
              >
                CONNECT
                <span
                  class="c12 c19 icon icon-caret-down "
                  color="text.secondary"
                  font-size="2"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
